### PR TITLE
Update: Behavior of --init (fixes #11105)

### LINKED ISSUE #11105: Update --init behavior
**The problem you want to solve.**

The `--init` behavior has remained pretty consistent since it was first launched. At that point, we didn't know what options people wanted and the number of rules was very small. It seems like the current questions don't make sense for how ESLint is used today. For example, forcing people to pick a style guide doesn't make sense for everyone and sends a message that ESLint is used primarily for styling. I'd like to update the behavior of `--init` to better reflect current users expectations and usage.

**Your take on the correct solution to problem.**

These are the ways I'd like to change `--init`:

1. **Assume `ecmaVersion` is 2018.** We first asked which version of ECMAScript they were using when ES6 was just being implemented. At this point, I think it's fair to assume people always want the latest (they can change it in their config if not).
2. **Ask how people want to use ESLint.** Instead of asking if users want to configure a style guide, ask "How do you want to use ESLint?", with the following options :
  * To check syntax only
  * To check syntax and find problems
  * To check syntax, find problems, and enforce code style
3. **Ask for framework question to replace JSX/React question.** Instead of asking if people are using JSX and then if they are using React, I'd create one question that asks if they are using either React (which implies JSX) or Vue (because there's a helpful plugin for Vue and it's pretty popular now). 

Optional:

* **Remove autoconfig option.** I'm not sure how people feel about this, but I'm skeptical that this option is used all that much thanks to the popularity of shareable configs. I'm in favor of removing it if others agree but have no problem leaving it in if people feel it's still helpful. 

**Are you willing to submit a pull request to implement this change?**

Yes.

### PR DESCRIPTION
**Pull request by nzakas**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain:

Updates the behavior of `--init` per #11105

**What changes did you make? (Give an overview)**

Rearranged the questions that are asked using `--init`. We now assume the latest version of ECMAScript is desired.

Also, there was a bug where if you chose a style guide the resulting config would have nothing but the resulting style guide. I fixed that.

The majority of the work was in rearranging the order of the questions. That part is pretty hard to follow but I did my best to clean it up for this round of changes. If we decide to remove autoconfig, I'd do some deeper cleaning.

In addition to the unit tests, I manually tested that each path through the questionnaire works locally.

**Is there anything you'd like reviewers to focus on?**

Anything that I missed?

### CODE REVIEW

File: lib/config/config-initializer.js [hunk 1]
```diff
@@ -417,16 +429,62 @@ function askInstallModules(modules, packageJsonExists) {
...
+            ]
+        },
+        {
+            type: "list",
+            name: "moduleType",
+            message: "What type of modules does your project use?",
+            default: "esm",
+            choices: [
+                { name: "JavaScript modules", value: "esm" },
+                { name: "CommonJS", value: "commonjs" },
```
Comment (not-an-aardvark) [inline lib/config/config-initializer.js]: I suspect the wording here might be confusing, because many users aren't familiar with the term "CommonJS" or the distinction between the two module systems. Someone unfamiliar with modules might think, "well, I don't know what CommonJS means but I'm writing JavaScript, so I'm probably using JavaScript modules", when in fact most people using Node are probably using CommonJS. 

Could we add some additional context to this question? I'm not sure what the best wording would be, but maybe it would look something like this:

> * CommonJS (usually used if your code runs in Node.js)
> * JavaScript modules (usually used if ...)

Alternatively, maybe we could put this question after the browser/node question and change the wording accordingly.

File: lib/config/config-initializer.js [hunk 2]
```diff
@@ -417,16 +429,62 @@ function askInstallModules(modules, packageJsonExists) {
...
+            ]
+        },
+        {
+            type: "checkbox",
+            name: "env",
+            message: "Where does your code run?",
+            default: ["browser"],
+            choices: [
+                { name: "Browser", value: "browser" },
+                { name: "Node", value: "node" }
```
Comment (not-an-aardvark) [inline lib/config/config-initializer.js]: Should we add a "neither" option? (I suspect most ESLint setups probably run in either a browser or Node, but for the exceptional cases it might be nice to not add the wrong globals.)

File: lib/config/config-initializer.js [hunk 3]
```diff
@@ -417,16 +429,62 @@ function askInstallModules(modules, packageJsonExists) {
...
+            choices: [
+                { name: "JavaScript modules", value: "esm" },
+                { name: "CommonJS", value: "commonjs" },
+                { name: "None of these", value: "none" }
+            ]
+        },
+        {
+            type: "list",
+            name: "framework",
+            message: "Which framework does your project use?",
```
Comment (not-an-aardvark) [inline lib/config/config-initializer.js]: It's a minor thing, but this wording could be interpreted to mean that all projects should be using some framework (either React, Vue, or some other framework not listed here). I assume this isn't the intended interpretation, since not all projects use frameworks. What would you think about rewording this question to avoid that potential interpretation? For example:

> Are you using either of the following frameworks?

File: lib/config/config-initializer.js [hunk 4]
```diff
@@ -417,16 +429,62 @@ function askInstallModules(modules, packageJsonExists) {
 function promptUser() {
 
     return inquirer.prompt([
+        {
+            type: "list",
+            name: "purpose",
+            message: "How would you like to use ESLint?",
+            default: "problems",
+            choices: [
+                { name: "To check syntax only", value: "syntax" },
```
Comment (not-an-aardvark) [inline lib/config/config-initializer.js]: I think the "check syntax only" option is a fairly niche use case -- most projects would benefit from at least using `eslint:recommended`. Additionally, some users might not understand what "check syntax" means (it means "only report syntax errors" rather than "report bug-prone syntactic patterns"), or they might not have a clear understanding of how syntax errors differ from semantic errors.

In your view, is this preference sufficiently common that it's worth having an option for? I don't feel very strongly about it, but I'd prefer either:

* Removing it from the `--init` options, or
* Moving it to the bottom of the three options here (if we wanted to retain ordering consistency, we could also just reverse the order of these options entirely).

File: lib/config/config-initializer.js [hunk 5]
```diff
@@ -239,43 +242,52 @@ function configureRules(answers, config) {
...
-        config.rules["linebreak-style"] = ["error", answers.linebreak];
-        config.rules.semi = ["error", answers.semi ? "always" : "never"];
+    // add in library information
+    if (answers.framework === "react") {
+        config.parserOptions.ecmaFeatures = {
+            jsx: true
+        };
+        config.plugins = ["react"];
+    } else if (answers.framework === "vue") {
+        config.plugins = ["vue"];
```
Comment (g-plane) [inline lib/config/config-initializer.js]: There is a special parser for Vue.js. (Also don't forget to update the function `getModulesList`.)
```suggestion
        config.parser = "vue-eslint-parser";
        config.plugins = ["vue"];
```

File: lib/config/config-initializer.js (hunk 1 above)
Comment (nzakas) [reply lib/config/config-initializer.js]: I'm not sure that this is any less clear than what we had previously. Prior to this change, there were two questions:

1. Do you use CommonJS?
1. Do you use ES6 modules?

IMHO, it was more confusing before because we were not presenting the two options as related to each other. I worry about adding too many words to these options lest people stop reading.

File: lib/config/config-initializer.js (hunk 2 above)
Comment (nzakas) [reply lib/config/config-initializer.js]: The choices here are represented as a checkboxes in the UI, so you can select one, both, or none. (This is unchanged from what we currently have.)

File: lib/config/config-initializer.js (hunk 3 above)
Comment (nzakas) [reply lib/config/config-initializer.js]: I think the third option, "None of these", makes this question clear. I'm not a fan of changing to a yes/no question and then not asking for a yes/no response.

File: lib/config/config-initializer.js (hunk 4 above)
Comment (nzakas) [reply lib/config/config-initializer.js]: We got some direct feedback a while ago that it wasn't clear ESLint could be used just for syntax checking. While it may be a niche case, I don't think it hurts anything to include it here. This way, new users can easily see that ESLint can be used for a number of different purposes.

I don't see the purpose of reversing the order of the options. Right now, there is a logical progression of options that include more functionality, and either way the common case is in the middle and the default.

As it is, the default selection in this menu is to check syntax and find problems

File: lib/config/config-initializer.js (hunk 5 above)
Comment (nzakas) [reply lib/config/config-initializer.js]: Is the Vue parser required for working Vue? Or is it optional?

Comment (g-plane) [reply lib/config/config-initializer.js]: It's required.

File: lib/config/config-initializer.js (hunk 1 above)
Comment (not-an-aardvark) [reply lib/config/config-initializer.js]: Agreed that asking two questions/adding too many words is not ideal. What about something like this?

> Are you using `import`/`export` syntax?

We would remove the explicit question about CommonJS, and instead enable the `commonjs` env when the user says they're both (a) using Node and (b) not using `import`/`export` syntax.

File: lib/config/config-initializer.js (hunk 5 above)
Comment (nzakas) [reply lib/config/config-initializer.js]: Okay, thank you. I will add that in.

File: lib/config/config-initializer.js (hunk 1 above)
Comment (nzakas) [reply lib/config/config-initializer.js]: We need to explicitly ask about CommonJS because it can be used in the browser, so we can't logically infer that if they're not using Node.js, then they must not be using CommonJS.

I'm not opposed to changing "JavaScript modules" to "import/export syntax", though, if you feel that makes it clearer.

File: lib/config/config-initializer.js (hunk 5 above)
Comment (mysticatea) [reply lib/config/config-initializer.js]: Actually, `eslint-plugin-vue` contains `vue-eslint-parser` as a dependency and it requires some additional configuration to work properly. The plugin provides `plugin:vue/base` preset for this.

> https://github.com/vuejs/eslint-plugin-vue/blob/f46c75b391422a651711fe67aac60963faf39406/lib/configs/base.js

- `parser: require.resolve('vue-eslint-parser'),` supports parsing of `*.vue` files.
- `'vue/comment-directive': 'error',` supports `<!-- eslint-disable -->` comment directives in Vue.js template.
- `parserOptions.ecmaFeatures.jsx: true` and `'vue/jsx-uses-vars': 'error'` supports JSX syntax.

Then, they provides `plugin:vue/essential` preset to catch problems in Vue.js files as smilar to `eslint:recommended`.

> https://github.com/vuejs/eslint-plugin-vue/blob/f46c75b391422a651711fe67aac60963faf39406/lib/configs/essential.js

Those rules catch only semantic errors.

Therefore, I'd like to recommend to use `extends: "plugin:vue/essential"` instead.

```suggestion
        config.extends = ["plugin:vue/essential"];
```

But if you don't want to contain rules which catch semantic errors, `plugin:vue/base` (configuration to parse `*.vue` files properly) is a good option.

Comment (nzakas) [reply lib/config/config-initializer.js]: Sounds good, I'll make these changes.

**Comment by nzakas** [comment]
@g-plane @mysticatea can you take a look at the latest changes to make sure they're correct?

File: lib/config/config-initializer.js [hunk 6]
```diff
@@ -239,43 +242,61 @@ function configureRules(answers, config) {
...
-    config.parserOptions.ecmaVersion = answers.ecmaVersion;
-    if (answers.ecmaVersion >= 2015) {
-        if (answers.modules) {
-            config.parserOptions.sourceType = "module";
-        }
-        config.env.es6 = true;
-    }
+    // set the latest ECMAScript version
+    config.parserOptions.ecmaVersion = DEFAULT_ECMA_VERSION;
+    config.env.es6 = true;
```
Comment (mysticatea) [inline lib/config/config-initializer.js]: ES2017 has added two globals: `Atomics` and `SharedArrayBuffer`. Because now `DEFAULT_ECMA_VERSION` is `2018` and `env.es6` doesn't include those, should we add `globals.Atomics` and `globals.SharedArrayBuffer` here?

File: lib/config/config-initializer.js (hunk 5 above)
Comment (mysticatea) [reply lib/config/config-initializer.js]: Thank you! LGTM about Vue's setting.

File: lib/config/config-initializer.js (hunk 6 above)
Comment (nzakas) [reply lib/config/config-initializer.js]: Oh yeah, that's a good idea. Thanks.

**Comment by nzakas** [comment]
Okay, I believe all of the outstanding issues have been addressed. Please check on this one more time. :)

File: lib/config/config-initializer.js [hunk 7]
```diff
@@ -239,43 +242,65 @@ function configureRules(answers, config) {
...
+    config.globals = {
+        Atomics: false,
```
Comment (platinumazure) [inline lib/config/config-initializer.js]: Now that #11338 is merged, should we use "readable"?

```suggestion
        Atomics: "readable",
```

File: lib/config/config-initializer.js [hunk 8]
```diff
@@ -239,43 +242,65 @@ function configureRules(answers, config) {
...
+        SharedArrayBuffer: false
```
Comment (platinumazure) [inline lib/config/config-initializer.js]: Now that #11338 is merged, should we use "readable"?

```suggestion
        SharedArrayBuffer: "readable"
```

**Comment by nzakas** [comment]
Once https://github.com/eslint/eslint/pull/11384 is merged, I'll update to use "readonly". Then this will be ready to merge.

File: lib/config/config-initializer.js (hunk 8 above)
Comment (nzakas) [reply lib/config/config-initializer.js]: Updating this to use "readonly"

**Comment by nzakas** [comment]
Okay, I've updated to use "readonly" instead of `false` in the generated config. Please take one last look and feel free to merge if there are no further concerns.
